### PR TITLE
Type hints added for python/ray/autoscaler/node_provider.py

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by yiranwang52** [comment]
Can you add a mypy check to CI?

File: python/ray/autoscaler/node_provider.py [hunk 1]
```diff
@@ -11,22 +11,22 @@
 logger = logging.getLogger(__name__)
 
 
-def _import_aws(provider_config):
+def _import_aws(provider_config: Dict[str, Any]) -> Any:
```
Comment (ericl) [inline python/ray/autoscaler/node_provider.py:14]:
```suggestion
def _import_aws(provider_config: Dict[str, Any]) -> "NodeProvider":
```

File: python/ray/autoscaler/node_provider.py [hunk 2]
```diff
@@ -118,7 +118,7 @@ def _import_external(provider_config):
 }
 
 
-def _load_class(path):
+def _load_class(path: str) -> Any:
```
Comment (ericl) [inline python/ray/autoscaler/node_provider.py:121]:
```suggestion
def _load_class(path: str) -> type:
```

File: python/ray/autoscaler/node_provider.py [hunk 3]
```diff
@@ -134,7 +134,7 @@ def _load_class(path):
 
 
 def _get_node_provider(provider_config: Dict[str, Any],
-                       cluster_name: str) -> Any:
+                       cluster_name: str) -> Union[Any, 'NotImplemented']:
```
Comment (ericl) [inline python/ray/autoscaler/node_provider.py:137]:
```suggestion
                       cluster_name: str) -> Union[Any, "NotImplemented"]:
```

**Review by ericl — CHANGES_REQUESTED** [review]
Thanks for looking at this. Some requested changes:

File: python/ray/autoscaler/node_provider.py (hunk 2 above)
Comment (Gekho457) [reply python/ray/autoscaler/node_provider.py:121]: Not sure if this is the right place for this, but will ask anyway:
Can you explain how the annotation "type" differs from "Any"?

File: python/ray/autoscaler/node_provider.py (hunk 3 above)
Comment (Gekho457) [reply python/ray/autoscaler/node_provider.py:137]: How would you recommend handling the methods which are not implemented yet, e.g. NodeProvider.is_running()?
https://github.com/ray-project/ray/pull/11023/files#diff-c732efe3edd32eb2ec55d17741db1c12L189

**Pull request closed without merging**